### PR TITLE
reafactor: add fsm, additional state handling

### PR DESCRIPTION
**Pull request by Schille**

### CODE REVIEW

File: client/beiboot/api.py [hunk 1]
```diff
@@ -139,6 +139,7 @@ def get_connection(
             # if the kubeconfig was added, this cluster is ready
             break
         else:
+            print(bbt)
```
Comment (buschNT) [inline client/beiboot/api.py:142]: remove?

File: operator/beiboot/clusterstate.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    The body of the Beiboot CRD is available as self.model
+    """
+
+    requested = State("Cluster requested", initial=True, value="REQUESTED")
+    creating = AsyncState("Cluster creating", value="CREATING")
+    pending = AsyncState("Cluster pending", value="PENDING")
+    running = AsyncState("Cluster running", value="RUNNING")
+    ready = AsyncState("Cluster running", value="READY")
+    error = AsyncState("Cluster error", value="ERROR")
+    terminating = AsyncState("Cluster terminating", value="TERMINATING")
```
Comment (buschNT) [inline operator/beiboot/clusterstate.py:32]: maybe "terminated", as this is the last state and the termination should be finished, or is this the state before reaching "terminated"?

Comment (Schille) [reply operator/beiboot/clusterstate.py:32]: "Terminated" means this Beiboot cluster (CRD) does not exist anymore in K8s

File: operator/beiboot/provider/k3s/__init__.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+from .utils import (
+    create_k3s_server_workload,
+    create_k3s_agent_workload,
+    create_k3s_kubeapi_service,
+)
+
+core_api = k8s.client.CoreV1Api()
+
+
+class K3s(AbstractClusterProvider):
```
Comment (buschNT) [inline operator/beiboot/provider/k3s/__init__.py:21]: Seems like we are duplicating code from getdeck here. Is this intended? Could make this hard to maintain.

File: operator/beiboot/handler/clusters.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+async def handle_cluster_workload_events(event, namespace, logger, **kwargs):
+    beiboot = get_beiboot_for_namespace(namespace)
+    if beiboot is None:
+        logger.warn(f"The Beiboot object for namespace '{namespace}' does not exist")
+        return
+    parameters = configuration.refresh_k8s_config()
+    cluster = BeibootCluster(configuration, parameters, model=beiboot, logger=logger)
+    # drop events that have been prior to last state change of cluster
+
+    if (
```
Comment (buschNT) [inline operator/beiboot/handler/clusters.py:55]: else missing?

File: operator/beiboot/utils.py [hunk 5]
```diff
@@ -224,3 +148,76 @@ def get_taken_gefyra_ports(
 def get_namespace_name(name: str, cluster_config: ClusterConfiguration) -> str:
     namespace = f"{cluster_config.namespacePrefix}-{name}"
     return namespace
+
+
+class AsyncTransition(Transition):
```
Comment (buschNT) [inline operator/beiboot/utils.py:153]: this part is hard to understand for me, maybe because I never worked with the statemachine package.

**Review by buschNT — APPROVED** [review]
I am not sure about the "ClusterFactory" pattern as we duplicate lots of code between getdeck and beiboot. maybe this is temporary? Can we remove this after splitting getdeck into a library and CLI? The library could then be used in beiboot (but I guess we would get a circular dependency since getdeck imports beiboot?).

File: client/beiboot/api.py [hunk 6]
```diff
@@ -79,11 +79,39 @@ def create_cluster(
...
 
-    if connect:
-        logger.info(
-            "Now connecting to the cluster; this may take a while to complete. "
-        )
-        establish_connection(cluster_name, probe_connection, configuration)
+    w = k8s.watch.Watch()
+
+    # wait for cluster events side is ready
+    for _object in w.stream(
```
Comment (SteinRobert) [inline client/beiboot/api.py:85]: Just wondering - would it make sense to add some kind of timeout here?

File: operator/beiboot/handler/beiboots.py [hunk 7]
```diff
@@ -1,266 +1,104 @@
...
-        if additional_services:
-            services.extend(additional_services)
-        #
-        # Gefyra Nodeport
-        #
-        if hasattr(cluster_config, "gefyra") and cluster_config.gefyra.get("enabled"):
+            await cluster.impair(str(e))
+            raise kopf.PermanentError(str(e))
+
+    if cluster.is_pending:
```
Comment (SteinRobert) [inline operator/beiboot/handler/beiboots.py:40]: Why not `elif`?

File: operator/beiboot/handler/beiboots.py [hunk 8]
```diff
@@ -1,266 +1,104 @@
...
+        try:
+            await cluster.operate()  # become running
+            await sleep(1)
+        except kopf.PermanentError as e:
+            await cluster.impair(str(e))
+            raise e from None
+    if cluster.is_running or (cluster.is_ready and await cluster.kubeconfig):
```
Comment (SteinRobert) [inline operator/beiboot/handler/beiboots.py:47]: Maybe `elif` too?

File: operator/beiboot/handler/clusters.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            # need to wrap this for correct later use
+            return AttrDict(bbt)
+    else:
+        return None
+
+
+@kopf.on.event(
+    "event",
+    when=kopf.all_(
+        [lambda event, **_: event["type"] is not None, workloads_in_beiboot_namespace]
```
Comment (SteinRobert) [inline operator/beiboot/handler/clusters.py:53]: Just style - why one lambda and one def - maybe both def is a bit more consistent. Took me a while to understand.

File: operator/beiboot/handler/clusters.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+import kopf
+
+from beiboot.clusterstate import BeibootCluster
+from beiboot.configuration import configuration
+
+core_api = k8s.client.CoreV1Api()
+objects_api = k8s.client.CustomObjectsApi()
+
+
+def _in_any_beiboot_namespace(event, namespace, kind: list, **_):
```
Comment (SteinRobert) [inline operator/beiboot/handler/clusters.py:13]: `kind` is a list should be plural.

**Review by SteinRobert — CHANGES_REQUESTED** [review]

File: operator/beiboot/handler/beiboots.py (hunk 7 above)
Comment (Schille) [reply operator/beiboot/handler/beiboots.py:40]: if the cluster was first in requested or creating state I want it to fall in this branch afterwards.

**Pull request merged**